### PR TITLE
Added GenerateCode and GenerateCodeCustom functions to both HOTP and TOTP

### LINKED ISSUE #5: Hiding Code Generation Makes Testing Hard
Hey @pquerna, thanks for this library! I'm currently reimplementing a security system from Python in Go that uses TOTP to control access to a physical location, and it's great to have a sane interface for validating codes.

The problem though, is that this library is *only* for server-side authentication and not for client-side authcode generation. This seems like a missed opportunity generally, but specifically it makes testing very awkward. I'd like to be able to generate a few keys for testing, then run my system through its paces using these keys, but the interface expects that external code will do the code generation.

Any way you could add a pair of generator methods to `totp.Key`, such as `CodeFromCounter(i int) string` and `CodeFromTime(t time.Time) string`, for testing and client generation uses?

Thanks!

### PR DESCRIPTION
**Pull request by cathalgarvey**
Closes #5

### CODE REVIEW

File: hotp/hotp.go [hunk 1]
```diff
@@ -57,27 +57,23 @@ type ValidateOpts struct {
...
-		if len(passcode) != 8 {
-			return false, otp.ErrValidateInputInvalidLength8
-		}
-	default:
-		panic("unsupported Digits value.")
-	}
+// GenerateCode creates a HOTP passcode given a counter and secret.
+// This is a shortcut for GenerateCodeCustom, with parameters that
+// are compataible with Google-Authenticator.
+func GenerateCode(secret string, counter uint64) string {
```
Comment (pquerna) [inline hotp/hotp.go]: Not returning error here is the only thing i'm not sure about.

For the simple `Validate` its a little easier to reason about only returning a `bool` and no `error` -- you only care if the  token was valid from a flow perspective, if it was malformed, or whatever, you want to show the user the invalid token flow -- for Generate, if the secret is malformed, returning an empty string.... is pretty rough.  So i'm thinking it's return signature should be `(string, error)`

**Comment by pquerna** [comment]
thank you for the PR, it looks great.  Just not sure about `Generate` return value right now.  Thoughts?

**Comment by cathalgarvey** [comment]
I'll fix it to return the error, no problem. Was just being consistent with the validate sigs but I see your point. :)

On 9 April 2016 16:59:23 IST, Paul Querna <notifications@github.com> wrote:
>thank you for the PR, it looks great.  Just not sure about `Generate`
>return value right now.  Thoughts?
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/pquerna/otp/pull/6#issuecomment-207811811

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cathalgarvey** [comment]
Sorry for delay, changes pushed. :)

**Comment by pquerna** [comment]
Thank you!

**Pull request merged**